### PR TITLE
feat: [SSCA-32]: SSCA orchestration step in new YAML format (beta)

### PR DESCRIPTION
**Pull request by udayvunnam**
### Summary

TODO: Update enums once BE changes are merged with updated swagger for the step

image

#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by udayvunnam** [comment]
retrigger titlecheck

**Comment by udayvunnam** [comment]
retrigger coverage

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+export interface SscaOrchestrationStepSpecUI {
+  sbomToolType?: MultiTypeSelectOption
+  sbomFormat?: MultiTypeSelectOption
+  sbomSourceArtfifactType?: MultiTypeSelectOption
+  attestationPrivateKey?: string
+  limitMemory?: string
+  limitCPU?: string
+}
+
+export interface SscaOrchestrationStepDataUI extends Omit<SscaOrchestrationStepData, 'spec'> {
```
Comment (chetanvekariya-harness) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx:57]: Omit might not be required. We can directly override spec, I guess.

**Review by chetanvekariya-harness — APPROVED** [review]

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepVariables.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  metadataMap: Required<VariableMergeServiceResponse>['metadataMap']
+  variablesData: SscaOrchestrationStepData
+}
+
+export const SscaOrchestrationStepVariables: React.FC<SscaOrchestrationStepVariablesProps> = ({
+  variablesData,
+  metadataMap,
+  initialValues
+}) => (
+  <VariablesListTable
```
[the remark this answers is not in the document]
Comment (reetika04) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepVariables.tsx:28]: +1

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepFunctionConfigs.tsx [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    label: 'connectors.serviceNow.privateKey',
+    isRequired: true
+  },
+  {
+    name: 'timeout',
+    type: ValidationFieldTypes.Timeout
+  }
+]
+
+export function getInputSetViewValidateFieldsConfig(
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepFunctionConfigs.tsx:85]: Where is this being used? Will isRequired ever be false?

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+export const artifactTypeOptions: SelectOption[] = [{ label: 'Image', value: 'image' }]
+
+export const sbomGenerationToolOptions: SelectOption[] = [{ label: 'Syft', value: 'Syft' }]
+
+export const sbomFormatOptions = [
+  { label: 'SPDX', value: 'SPDX' },
+  { label: 'CycloneDX', value: 'CycloneDX' }
+]
+
+const SscaOrchestrationStepBase = (
```
[the remark this answers is not in the document]
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx:41]: This naming convention is similar to CI Steps as this was initially created as CI step and followed the same conventions

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepFunctionConfigs.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+ * Copyright 2021 Harness Inc. All rights reserved.
+ * Use of this source code is governed by the PolyForm Shield 1.0.0 license
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
+ */
+
+import { Types as TransformValuesTypes } from '@pipeline/components/PipelineSteps/Steps/StepsTransformValuesUtils'
+import { Types as ValidationFieldTypes } from '@pipeline/components/PipelineSteps/Steps/StepsValidateUtils'
+
+export const transformValuesFieldsConfig = [
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepFunctionConfigs.tsx:11]: Lot of fields that are common here as well which can be abstracted out.

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+export class SscaOrchestrationStep extends PipelineStep<SscaOrchestrationStepData> {
+  constructor() {
+    super()
+    this._hasStepVariables = true
+  }
+
+  protected type = StepType.SscaOrchestration
+  protected stepName = 'Ssca Orchestration'
+  protected stepIcon: IconName = 'ssca-orchestrate'
+  protected stepIconColor = '#4F5162'
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx:83]: This is Color.GREY_600

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+  }
+  attestation: {
+    privateKey: string
+  }
+}
+
+export interface SscaOrchestrationStepData {
+  name?: string
+  identifier: string
+  type: string
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx:43]: Type should be an enum

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+import { SscaOrchestrationStepBaseWithRef } from './SscaOrchestrationStepBase'
+import { transformValuesFieldsConfig } from './SscaOrchestrationStepFunctionConfigs'
+import SscaOrchestrationStepInputSet from './SscaOrchestrationStepInputSet'
+import { SscaOrchestrationStepVariables, SscaOrchestrationStepVariablesProps } from './SscaOrchestrationStepVariables'
+
+export interface SscaOrchestrationStepSpec {
+  tool: {
+    type: string
+    spec: {
+      format: string
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx:29]: enum

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    return getFormValuesInCorrectFormat<T, SscaOrchestrationStepData>(data, transformValuesFieldsConfig)
+  }
+
+  validateInputSet({
+    data,
+    template,
+    getString,
+    viewType
+  }: ValidateInputSetProps<SscaOrchestrationStepData>): FormikErrors<SscaOrchestrationStepData> {
+    const errors = {} as any
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx:115]: as any?

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+import type { BuildStageElementConfig } from '@pipeline/utils/pipelineTypes'
+import type {
+  SscaOrchestrationStepData,
+  SscaOrchestrationStepDataUI,
+  SscaOrchestrationStepProps
+} from './SscaOrchestrationStep'
+import { editViewValidateFieldsConfig, transformValuesFieldsConfig } from './SscaOrchestrationStepFunctionConfigs'
+import css from './SscaOrchestrationStep.module.scss'
+
+export const artifactTypeOptions: SelectOption[] = [{ label: 'Image', value: 'image' }]
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx:32]: Use strings for labels

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+export const sbomGenerationToolOptions: SelectOption[] = [{ label: 'Syft', value: 'Syft' }]
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx:34]: Use strings for labels

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+export const sbomFormatOptions = [
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx:36]: Use strings for labels

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+              <FormInput.Select
+                items={artifactTypeOptions}
+                name={`spec.source.type`}
+                label={getString('pipeline.artifactsSelection.artifactType')}
+                placeholder={getString('select')}
+                disabled={readonly}
+              />
+
+              <Text font={{ variation: FontVariation.LEAD }} color={Color.GREY_900} margin={{ top: 'medium' }}>
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx:143]: font={{ variation: FontVariation.FORM_LABEL }}

Comment (reetika04) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx:143]: Why aren’t we using that?

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx (hunk 8 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx:29]: For a couple of things we are yet to get enum values from BE, they will be generated once the API/swagger is ready

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx (hunk 13 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx:143]: This is not form label. Check the attached screen shot. More apt would be FORM_SUB_SECTION/LEAD

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx (hunk 12 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx:36]: All of these will be replaced by the enums once BE swagger is integrated

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx (hunk 1 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx:57]: Types of property 'spec' are incompatible. They need to have similar structure unless

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx:41]: Done

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepFunctionConfigs.tsx (hunk 3 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepFunctionConfigs.tsx:85]: Please check the latest commit

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+import type { BuildStageElementConfig } from '@pipeline/utils/pipelineTypes'
+import type {
+  SscaOrchestrationStepData,
+  SscaOrchestrationStepDataUI,
+  SscaOrchestrationStepProps
+} from './SscaOrchestrationStep'
+import { editViewValidateFieldsConfig, transformValuesFieldsConfig } from './SscaOrchestrationStepFunctionConfigs'
+import css from './SscaOrchestrationStep.module.scss'
+
+export const artifactTypeOptions: SelectOption[] = [{ label: 'Image', value: 'image' }]
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx:32]: Labels should be strings.

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+
+export const sbomGenerationToolOptions: SelectOption[] = [{ label: 'Syft', value: 'Syft' }]
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx:34]: Labels should be strings.

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+
+export const sbomFormatOptions = [
```
Comment (reetika04) [inline src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx:36]: Labels should be strings.

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx (hunk 11 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepBase.tsx]: @Reetika These are constants outside the component where we don't have getString. Besides these are not language specific to move to strings yaml

**Comment by udayvunnam** [comment]
retrigger jest

**Review by sonaligoyal-harness — APPROVED** [review]

**Comment by udayvunnam** [comment]
Constants outside the component are not kept in strings yaml as they are already constants. Besides these are not language specific to move to strings yaml

**Review by bhavya-sinha — APPROVED** [review]

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx (hunk 6 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx]: done

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepFunctionConfigs.tsx (hunk 5 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepFunctionConfigs.tsx:11]: This followed the CI step structure. Agreed we can abstract it out. Will try in a different PR

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx (hunk 16 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx:36]: https://github.com/harness/harness-core-ui/pull/14312#issuecomment-1448204608

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx (hunk 15 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx:34]: https://github.com/harness/harness-core-ui/pull/14312#issuecomment-1448204608

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx (hunk 14 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStepEdit.tsx:32]: https://github.com/harness/harness-core-ui/pull/14312#issuecomment-1448204608

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+import type { MultiTypeSelectOption } from '@pipeline/components/PipelineSteps/Steps/StepsTypes'
+import type { StringsMap } from 'stringTypes'
+import { SscaOrchestrationStepBaseWithRef } from './SscaOrchestrationStepBase'
+import { transformValuesFieldsConfig } from './SscaOrchestrationStepFunctionConfigs'
+import SscaOrchestrationStepInputSet from './SscaOrchestrationStepInputSet'
+import { SscaOrchestrationStepVariables, SscaOrchestrationStepVariablesProps } from './SscaOrchestrationStepVariables'
+
+export interface SscaOrchestrationStepSpec {
+  tool: {
+    type: string
```
[the remark this answers is not in the document]
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx:32]: https://github.com/harness/harness-core-ui/pull/14312#discussion_r1119662669

File: src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx (hunk 7 above)
Comment (udayvunnam) [reply src/modules/75-ssca/components/PipelineSteps/SscaOrchestrationStep/SscaOrchestrationStep.tsx:48]: https://github.com/harness/harness-core-ui/pull/14312#discussion_r1119662669

**Review by vkbansal-harness — APPROVED** [review]

**Pull request merged**